tutorials/getting_started/semantic_parsing.md

```diff
@@ -0,0 +1,315 @@
+
+```python
+class TransitionFunction(torch.nn.Module, Generic[StateType]):
+    def take_step(self,
+                  state: StateType,
+                  max_actions: int = None,
+                  allowed_actions: List[Set] = None) -> List[StateType]:
+        raise NotImplementedError
+```
+
+If you ignore the optional arguments for a minute, this is a simple function that takes a `State`
+of a particular type and returns a list of `States` of the same type.  This function will have some
+parameters, probably based on some kind of LSTM decoder hidden state, that are used to score each
+valid action, construct new `States` for those actions, and return them.  The optional parameters
+are for efficiency, so you don't construct huge numbers of `State` objects if you know you're doing
+a beam search with size `k`, or if you're training and only need to score a few select action
+sequences.
+
+The `TransitionFunction` (along with a `State` object which we'll get to later TODO: add link)
+forms the basis of our decoder state machine, and now we need some way to train it.  There are a
+lot of ways to train state machines, depending on what kind of supervision you have.  You could
+have fully-labeled correct action sequences, a set of possibly correct action sequences, a way to
+check the correctness of finished states, or just a reward function on finished or intermediate
+states.  To handle the variety of supervision signals that could be used, we provide a simple
+`DecoderTrainer` interface that's generic over the supervision type:
+
+```python
+class DecoderTrainer(Generic[SupervisionType]):
+    def decode(self,
+               initial_state: State,
+               transition_function: TransitionFunction,
+               supervision: SupervisionType) -> Dict[str, torch.Tensor]:
+        raise NotImplementedError
+```
+
+The `SupervisionType` here could be a correct action sequence or a reward function (or anything
+else you choose); the generic interface makes it so that you can easily swap out trainers that take
+the same supervision signal (such as maximum likelihood with beam search optimization, or various
+reinforcement learning algorithms with each other).  The `DecoderTrainer` requires an initial
+`State`, a `TransitionFunction`, and a supervision signal, and it returns a `loss` tensor that we
+use to compute gradients and train the `TransitionFunction` and the encoder that created the
+initial `State` (the return value is a dictionary so you can return whatever else you want, too).
+
+There are several examples of `DecoderTrainers` and `TransitionFunctions` already in the library,
+and you can create your own if what we have doesn't fit your needs (and if you implement your own
+that you think would be useful for others, please consider contributing it back!).  Here are a few
+examples (not an exhaustive list): TODO: add links
+
+`DecoderTrainers`:
+- MaximumMarginalLikelihood (when you have a set of possibly correct action sequences)
```

Comment:
Give concrete examples of where these are used. E.g wikitables, all action sequences which execute to the correct logical form, ERM for NLVR etc.